File: concept/thing/ThingImpl.ts
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {Thing as ThingProto, Type as TypeProto} from "typedb-protocol/common/concept_pb";
import {Transaction as TransactionProto} from "typedb-protocol/common/transaction_pb";
import {Concept} from "../../api/concept/Concept";
import {Attribute} from "../../api/concept/thing/Attribute";
import {Relation} from "../../api/concept/thing/Relation";
import {Thing} from "../../api/concept/thing/Thing";
import {AttributeType} from "../../api/concept/type/AttributeType";
import {RoleType} from "../../api/concept/type/RoleType";
import {ThingType} from "../../api/concept/type/ThingType";
import {TypeDBTransaction} from "../../api/connection/TypeDBTransaction";
import {ErrorMessage} from "../../common/errors/ErrorMessage";
import {TypeDBClientError} from "../../common/errors/TypeDBClientError";
import {RequestBuilder} from "../../common/rpc/RequestBuilder";
import {Stream} from "../../common/util/Stream";
import {
    AttributeImpl,
    AttributeTypeImpl,
    ConceptImpl,
    EntityImpl,
    RelationImpl,
    RoleTypeImpl
} from "../../dependencies_internal";
import Annotation = ThingType.Annotation;
import BAD_ENCODING = ErrorMessage.Concept.BAD_ENCODING;
import BAD_VALUE_TYPE = ErrorMessage.Concept.BAD_VALUE_TYPE;

export abstract class ThingImpl extends ConceptImpl implements Thing {
    private readonly _iid: string;
    private readonly _inferred: boolean;

    protected constructor(iid: string, inferred: boolean) {
        super();
        if (!iid) throw new TypeDBClientError(ErrorMessage.Concept.MISSING_IID);
        this._iid = iid;
        this._inferred = inferred;
    }

    abstract asRemote(transaction: TypeDBTransaction): Thing.Remote;

    equals(concept: Concept): boolean {
        if (!concept.isThing()) return false;
        else return concept.asThing().iid === this._iid;
    }

    toString(): string {
        return `${this.className}[iid:${this._iid}]`;
    }

    get iid(): string {
        return this._iid;
    }

    abstract get type(): ThingType;

    get inferred(): boolean {
        return this._inferred;
    }

    isThing(): boolean {
        return true;
    }

    asThing(): Thing {
        return this;
    }

    toJSONRecord(): Record<string, boolean | string | number> {
        return {type: this.type.label.name};
    }
}

export namespace ThingImpl {

    export abstract class Remote extends ConceptImpl.Remote implements Thing.Remote {

        private readonly _iid: string;
        private readonly _inferred: boolean;

        protected constructor(transaction: TypeDBTransaction.Extended, iid: string, inferred: boolean, ..._: any) {
            super(transaction);
            if (!iid) throw new TypeDBClientError(ErrorMessage.Concept.MISSING_IID);
            this._iid = iid;
            this._inferred = inferred;
        }

        abstract asRemote(transaction: TypeDBTransaction): Thing.Remote;

        equals(concept: Concept): boolean {
            if (concept.isType()) return false;
            else return concept.asThing().iid === this._iid;
        }

        toString(): string {
            return `${this.className}[iid:${this._iid}]`;
        }

        get iid(): string {
            return this._iid;
        }

        abstract get type(): ThingType;

        get inferred(): boolean {
            return this._inferred;
        }

        isThing(): boolean {
            return true;
        }

        asThing(): Thing.Remote {
            return this;
        }

        toJSONRecord(): Record<string, boolean | string | number> {
            return {type: this.type.label.name};
        }

        async delete(): Promise<void> {
            const request = RequestBuilder.Thing.deleteReq(this.iid);
            await this.execute(request);
        }

        getHas(): Stream<Attribute>;
        getHas(annotations: Annotation[]): Stream<Attribute>;
        getHas(attributeType: AttributeType.Boolean): Stream<Attribute.Boolean>;
        getHas(attributeType: AttributeType.Long): Stream<Attribute.Long>;
        getHas(attributeType: AttributeType.Double): Stream<Attribute.Double>;
        getHas(attributeType: AttributeType.String): Stream<Attribute.String>;
        getHas(attributeType: AttributeType.DateTime): Stream<Attribute.DateTime>;
        getHas(attributeTypes: AttributeType[]): Stream<Attribute>;
        getHas(annotationOrAttrTypeOrAttrTypes?: Annotation[] | AttributeType | AttributeType[]): Stream<Attribute> {
            let isSingleAttrType = false;
            let request;
            if (typeof annotationOrAttrTypeOrAttrTypes === "undefined") {
                request = RequestBuilder.Thing.getHasReqByAnnotations(this.iid, []);
            } else if (Array.isArray(annotationOrAttrTypeOrAttrTypes)) {
                const asArray = annotationOrAttrTypeOrAttrTypes as Array<any>;
                if (asArray.length == 0 || asArray[0] instanceof AttributeTypeImpl) {
                    request = RequestBuilder.Thing.getHasByTypeReq(
                        this.iid,
                        (annotationOrAttrTypeOrAttrTypes as AttributeType[]).map(attrType => ThingType.proto(attrType)));
                } else {
                    request = RequestBuilder.Thing.getHasReqByAnnotations(
                        this.iid,
                        (annotationOrAttrTypeOrAttrTypes as Annotation[]).map(annotation => Annotation.proto(annotation))
                    );
                }
            } else {
                request = RequestBuilder.Thing.getHasByTypeReq(this.iid, [ThingType.proto(annotationOrAttrTypeOrAttrTypes)]);
                isSingleAttrType = true;
            }
            const attributes = this.stream(request).flatMap(
                (resPart) => Stream.array(resPart.getThingGetHasResPart().getAttributesList())
            ).map((attrProto) => AttributeImpl.of(attrProto));
            if (isSingleAttrType) {
                const arg = annotationOrAttrTypeOrAttrTypes as AttributeType;
                if (arg.isBoolean()) return attributes as Stream<Attribute.Boolean>;
                else if (arg.isLong()) return attributes as Stream<Attribute.Long>;
                else if (arg.isDouble()) return attributes as Stream<Attribute.Double>;
                else if (arg.isString()) return attributes as Stream<Attribute.String>;
                else if (arg.isDateTime()) return attributes as Stream<Attribute.DateTime>;
                else throw new TypeDBClientError(BAD_VALUE_TYPE.message(arg));
            } else {
                return attributes;
            }
        }

        getPlaying(): Stream<RoleType> {
            const request = RequestBuilder.Thing.getPlayingReq(this.iid);
            return this.stream(request)
                .flatMap((resPart) => Stream.array(resPart.getThingGetPlayingResPart().getRoleTypesList()))
                .map((res) => RoleTypeImpl.of(res));
        }

        getRelations(roleTypes?: RoleType[]): Stream<Relation> {
            if (!roleTypes) roleTypes = [];
            const request = RequestBuilder.Thing.getRelationsReq(this.iid, roleTypes.map((roleType) => RoleType.proto(roleType)));
            return this.stream(request)
                .flatMap((resPart) => Stream.array(resPart.getThingGetRelationsResPart().getRelationsList()))
                .map((res) => RelationImpl.of(res));
        }

        async isDeleted(): Promise<boolean> {
            return !(await this.transaction.concepts.getThing(this.iid));
        }

        async setHas(attribute: Attribute): Promise<void> {
            const request = RequestBuilder.Thing.setHasReq(this.iid, Thing.proto(attribute));
            await this.execute(request);
        }

        async unsetHas(attribute: Attribute): Promise<void> {
            const request = RequestBuilder.Thing.unsetHasReq(this.iid, Thing.proto(attribute));
            await this.execute(request);
        }

        protected async execute(request: TransactionProto.Req): Promise<ThingProto.Res> {
            return (await this.transaction.rpcExecute(request, false)).getThingRes();
        }

        protected stream(request: TransactionProto.Req): Stream<ThingProto.ResPart> {
            return this.transaction.rpcStream(request).map((res) => res.getThingResPart());
        }
    }

    export function of(thingProto: ThingProto) {
        switch (thingProto.getType().getEncoding()) {
            case TypeProto.Encoding.ENTITY_TYPE:
                return EntityImpl.of(thingProto);
            case TypeProto.Encoding.RELATION_TYPE:
                return RelationImpl.of(thingProto);
            case TypeProto.Encoding.ATTRIBUTE_TYPE:
                return AttributeImpl.of(thingProto);
            default:
                throw new TypeDBClientError(BAD_ENCODING.message(thingProto.getType().getEncoding()));
        }
    }
}

File: concept/thing/ThingImpl.java
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.vaticle.typedb.client.concept.thing;

import com.vaticle.typedb.client.api.TypeDBTransaction;
import com.vaticle.typedb.client.api.concept.thing.Attribute;
import com.vaticle.typedb.client.api.concept.thing.Thing;
import com.vaticle.typedb.client.api.concept.type.AttributeType;
import com.vaticle.typedb.client.api.concept.type.RoleType;
import com.vaticle.typedb.client.common.exception.TypeDBClientException;
import com.vaticle.typedb.client.concept.ConceptImpl;
import com.vaticle.typedb.client.concept.type.RoleTypeImpl;
import com.vaticle.typedb.client.concept.type.ThingTypeImpl;
import com.vaticle.typedb.protocol.ConceptProto;
import com.vaticle.typedb.protocol.TransactionProto;
import com.vaticle.typeql.lang.common.TypeQLToken;

import java.util.Collections;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Stream;

import static com.vaticle.typedb.client.common.exception.ErrorMessage.Concept.BAD_ENCODING;
import static com.vaticle.typedb.client.common.exception.ErrorMessage.Concept.MISSING_IID;
import static com.vaticle.typedb.client.common.exception.ErrorMessage.Concept.MISSING_TRANSACTION;
import static com.vaticle.typedb.client.common.rpc.RequestBuilder.Thing.deleteReq;
import static com.vaticle.typedb.client.common.rpc.RequestBuilder.Thing.getHasReq;
import static com.vaticle.typedb.client.common.rpc.RequestBuilder.Thing.getPlayingReq;
import static com.vaticle.typedb.client.common.rpc.RequestBuilder.Thing.getRelationsReq;
import static com.vaticle.typedb.client.common.rpc.RequestBuilder.Thing.protoThing;
import static com.vaticle.typedb.client.common.rpc.RequestBuilder.Thing.setHasReq;
import static com.vaticle.typedb.client.common.rpc.RequestBuilder.Thing.unsetHasReq;
import static com.vaticle.typedb.client.concept.type.TypeImpl.protoAnnotations;
import static com.vaticle.typedb.client.concept.type.TypeImpl.protoTypes;
import static com.vaticle.typedb.common.util.Objects.className;
import static java.util.Arrays.asList;
import static java.util.Collections.emptySet;

public abstract class ThingImpl extends ConceptImpl implements Thing {

    private final String iid;
    private final boolean isInferred;

    ThingImpl(String iid, boolean isInferred) {
        if (iid == null || iid.isEmpty()) throw new TypeDBClientException(MISSING_IID);
        this.iid = iid;
        this.isInferred = isInferred;
    }

    public static ThingImpl of(ConceptProto.Thing thingProto) {
        switch (thingProto.getType().getEncoding()) {
            case ENTITY_TYPE:
                return EntityImpl.of(thingProto);
            case RELATION_TYPE:
                return RelationImpl.of(thingProto);
            case ATTRIBUTE_TYPE:
                return AttributeImpl.of(thingProto);
            case UNRECOGNIZED:
            default:
                throw new TypeDBClientException(BAD_ENCODING, thingProto.getType().getEncoding());
        }
    }

    @Override
    public final String getIID() {
        return iid;
    }

    @Override
    public abstract ThingTypeImpl getType();

    @Override
    public boolean isInferred() {
        return isInferred;
    }

    @Override
    public ThingImpl asThing() {
        return this;
    }

    @Override
    public String toString() {
        return className(this.getClass()) + "[" + getType().getLabel() + ":" + iid + "]";
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        ThingImpl that = (ThingImpl) o;
        return (this.iid.equals(that.iid));
    }

    @Override
    public int hashCode() {
        return iid.hashCode();
    }

    public abstract static class Remote extends ConceptImpl.Remote implements Thing.Remote {

        final TypeDBTransaction.Extended transactionRPC;
        private final String iid;
        private final boolean isInferred;
        private final int hash;

        Remote(TypeDBTransaction transaction, String iid, boolean isInferred) {
            if (transaction == null) throw new TypeDBClientException(MISSING_TRANSACTION);
            this.transactionRPC = (TypeDBTransaction.Extended) transaction;
            if (iid == null || iid.isEmpty()) throw new TypeDBClientException(MISSING_IID);
            this.iid = iid;
            this.isInferred = isInferred;
            this.hash = Objects.hash(this.transactionRPC, this.getIID());
        }

        @Override
        public final String getIID() {
            return iid;
        }

        @Override
        public abstract ThingTypeImpl getType();

        @Override
        public boolean isInferred() {
            return isInferred;
        }

        @Override
        public final Stream<AttributeImpl<?>> getHas(AttributeType... attributeTypes) {
            return stream(getHasReq(getIID(), protoTypes(asList(attributeTypes))))
                    .flatMap(rp -> rp.getThingGetHasResPart().getAttributesList().stream())
                    .map(AttributeImpl::of);
        }

        @Override
        public final Stream<AttributeImpl.Boolean> getHas(AttributeType.Boolean attributeType) {
            return getHas((AttributeType) attributeType).map(AttributeImpl::asBoolean);
        }

        @Override
        public final Stream<AttributeImpl.Long> getHas(AttributeType.Long attributeType) {
            return getHas((AttributeType) attributeType).map(AttributeImpl::asLong);
        }

        @Override
        public final Stream<AttributeImpl.Double> getHas(AttributeType.Double attributeType) {
            return getHas((AttributeType) attributeType).map(AttributeImpl::asDouble);
        }

        @Override
        public final Stream<AttributeImpl.String> getHas(AttributeType.String attributeType) {
            return getHas((AttributeType) attributeType).map(AttributeImpl::asString);
        }

        @Override
        public final Stream<AttributeImpl.DateTime> getHas(AttributeType.DateTime attributeType) {
            return getHas((AttributeType) attributeType).map(AttributeImpl::asDateTime);
        }

        @Override
        public Stream<? extends Attribute<?>> getHas() {
            return getHas(emptySet());
        }

        @Override
        public final Stream<AttributeImpl<?>> getHas(Set<TypeQLToken.Annotation> annotations) {
            return stream(getHasReq(getIID(), protoAnnotations(annotations)))
                    .flatMap(rp -> rp.getThingGetHasResPart().getAttributesList().stream())
                    .map(AttributeImpl::of);
        }

        @Override
        public final Stream<RelationImpl> getRelations(RoleType... roleTypes) {
            return stream(getRelationsReq(getIID(), protoTypes(asList(roleTypes))))
                    .flatMap(rp -> rp.getThingGetRelationsResPart().getRelationsList().stream())
                    .map(RelationImpl::of);
        }

        @Override
        public final Stream<RoleTypeImpl> getPlaying() {
            return stream(getPlayingReq(getIID()))
                    .flatMap(rp -> rp.getThingGetPlayingResPart().getRoleTypesList().stream())
                    .map(RoleTypeImpl::of);
        }

        @Override
        public final void setHas(Attribute<?> attribute) {
            execute(setHasReq(getIID(), protoThing(attribute.getIID())));
        }

        @Override
        public final void unsetHas(Attribute<?> attribute) {
            execute(unsetHasReq(getIID(), protoThing(attribute.getIID())));
        }

        @Override
        public final void delete() {
            execute(deleteReq(getIID()));
        }

        @Override
        public final boolean isDeleted() {
            return transactionRPC.concepts().getThing(getIID()) == null;
        }

        @Override
        public final ThingImpl.Remote asThing() {
            return this;
        }

        protected ConceptProto.Thing.Res execute(TransactionProto.Transaction.Req.Builder request) {
            return transactionRPC.execute(request).getThingRes();
        }

        protected Stream<ConceptProto.Thing.ResPart> stream(TransactionProto.Transaction.Req.Builder request) {
            return transactionRPC.stream(request).map(TransactionProto.Transaction.ResPart::getThingResPart);
        }

        @Override
        public String toString() {
            return className(this.getClass()) + "[iid:" + iid + "]";
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            ThingImpl.Remote that = (ThingImpl.Remote) o;
            return this.transactionRPC.equals(that.transactionRPC) && this.iid.equals(that.iid);
        }

        @Override
        public int hashCode() {
            return hash;
        }
    }
}
